### ISSUE TITLE
string.quote vs process.quote

### ISSUE THREAD

**Issue by smimram** [savonet/liquidsoap#1576]
Right now we have two functions for quoting strings: `string.quote` and `process.quote`. We should decide whether we want to keep the two or not, and in the second case we should make clear the difference between the two (I could not really understand that from the doc).

**Comment by dtcooper** [comment]
FYI: `process.quote(...)` was a suggestion I made and @toots implemented to fix a security hole, that `string.quote(...)` still suffers from. [See relevant discussion here](https://github.com/savonet/liquidsoap/issues/1215). In my opinion we should remove (or alias) `string.quote`. Thanks!

**Comment by smimram** [comment]
Thanks @dtcooper for the explanation. I find this both confusing and worrying since the stnadard library uses `string.quote` where some process execution is involved (and thus `process.quote` should be used instead if I understand correctly):
```
$ git grep string.quote
deprecations.liq:# Deprecated: this function has been replaced by `string.quote`.
deprecations.liq:  deprecated("quote", "string.quote")
deprecations.liq:  string.quote(s)
externals.liq:    json = process.read("#{ffprobe} -print_format json -show_streams #{string.quote(fname)}")
externals.liq:          process.read.lines("#{mpcdec} -i #{string.quote(file)} 2>&1 \
externals.liq:    mpcdec_p = fun(f) -> "#{mpcdec} #{string.quote(f)} -"
externals.liq:                                              --show-channels #{string.quote(fname)}"))
externals.liq:                   #{string.quote(fname)}")
externals.liq:    faad_p = (fun (f) -> "#{faad} -w #{string.quote(f)}")
externals.liq:                           process.read.lines("#{faad} -i #{string.quote(file)} 2>&1 | \
externals.liq:        ret = process.read.lines("#{faad} -i #{string.quote(file)} 2>&1")
externals.liq:  osd = 'osd_cat -p #{position} --font #{string.quote(font)}'
externals.liq:    ignore(process.run("echo #{string.quote(display(m))} | #{osd} &"))
externals.liq:       ^ ' -t #{time} #{string.quote(title)} '
externals.liq:  s.on_metadata(fun (m) -> ignore(process.run(send^string.quote(display(m)))))
externals.liq:                       -vc null -vo null #{string.quote(s)} 2>/dev/null")
file.liq:      list.hd(default="", process.read.lines("file -b --mime-type #{string.quote(fname)}"))
http.liq:  list.hd(default="", process.read.lines("file -b -I #{string.quote(file)}"))
protocols.liq:    cmd = cmd % [("input",string.quote(input)),
protocols.liq:                 ("output",string.quote(output)),
protocols.liq:  cmd = "#{binary} --get-title --get-filename -- #{string.quote(arg)}"
protocols.liq:  cmd   = "rm -f $(output) && #{binary} -q -f best --no-playlist -o $(output) -- #{string.quote(arg)}"
protocols.liq:    ["annotate:title=#{string.quote(title)}:#{process}"]
protocols.liq:      "#{key}=#{string.quote(value)}"
protocols.liq:    cmd = "#{ffmpeg} -i #{string.quote(file)} -f ffmetadata - 2>/dev/null | grep -v '^;'"
protocols.liq:    cmd = "#{ffmpeg} -y -i $(input) -ac #{channels} #{cue_points} #{fades} #{replaygain_filter} #{string.quote(wav)}"
protocols.liq:  [process_uri(extname="wav", "echo #{string.quote(arg)} | #{binary} -scale 1.9 > $(output)")]
protocols.liq:  [process_uri(extname="mp3", "#{binary} -o $(output) #{string.quote(arg)}")]
protocols.liq:      "#{aws} --endpoint-url #{string.quote(endpoint)}"
protocols.liq:    "#{aws} --profile #{string.quote(profile)}"
protocols.liq:  cmd = "#{aws} --text #{string.quote(text)} --voice-id #{string.quote(voice_id)} $(output)"
resolvers.liq:  p = process.run("#{extract_replaygain} #{string.quote(file)}")
```
@toots what's your opinion?

**Comment by toots** [comment]
Just add a quick look:
* `string.quote` is a custom-made implementation to specifically escape `"` characters with a state-like implementation and double escapes
* `process.quote` is a OCaml-based implementation which includes OS-specific details and also probably benefits from a much greater exposure and, thus bug fixing.

Seems like I'd be all about deprecating `string.quote` indeed. I would also replace `Utils.quote` use by `Filename.quote`.

If everyone is down with the plan I'm happy to execute.

**Comment by smimram** [comment]
It looks fine by me. Having a custom quoting function is a notoriously bad idea...

**Comment by smimram** [comment]
I already have a problem with your fix. The `youtube-dl` protocol is implemented with
```
    ["annotate:title=#{string.quote(title)}:#{cmd}"]
```
Before, it was essentially producing a request like
```
annotate:title="bla":video.mp4
```
but now it is producing
```
annotate:title='bla':video.mp4
```
which is not valid. How do we proceed from there?

**Comment by smimram** [comment]
Fixed by ebb06c7a201d75e302c19aff56a80ee917267790.
